orca-kayenta/src/main/kotlin/com/netflix/spinnaker/orca/kayenta/tasks/PropagateDeployedClusterScopeTask.kt

```diff
@@ -16,46 +16,64 @@
 import com.netflix.spinnaker.orca.TaskResult
 import com.netflix.spinnaker.orca.ext.mapTo
-import com.netflix.spinnaker.orca.kayenta.model.CanaryConfigScope
-import com.netflix.spinnaker.orca.kayenta.model.KayentaCanaryContext
 import com.netflix.spinnaker.orca.kayenta.model.deployments
+import com.netflix.spinnaker.orca.kayenta.model.locations
+import com.netflix.spinnaker.orca.pipeline.model.Execution
 import com.netflix.spinnaker.orca.pipeline.model.Stage
 import org.springframework.stereotype.Component
 
 @Component
 class PropagateDeployedClusterScopeTask : Task {
   override fun execute(stage: Stage): TaskResult {
-    val (_, control, experiment) = stage.deployments
-    val canary = stage.parent?.mapTo<KayentaCanaryContext>("/canaryConfig")
-      ?: throw IllegalStateException("No parent stage found")
-
-    val deployedClusterScope = CanaryConfigScope(
-      controlScope = control.moniker.cluster,
-      controlLocation = control.region
-        ?: control.availabilityZones.keys.firstOrNull(),
-      experimentScope = experiment.moniker.cluster,
-      experimentLocation = experiment.region
-        ?: experiment.availabilityZones.keys.firstOrNull(),
-      startTimeIso = null,
-      endTimeIso = null,
-      extendedScopeParams = mapOf("type" to "cluster")
-    )
-
-    return if (canary.scopes.find {
-        it.controlScope == deployedClusterScope.controlScope &&
-          it.controlLocation == deployedClusterScope.controlLocation &&
-          it.experimentScope == deployedClusterScope.experimentScope &&
-          it.experimentLocation == deployedClusterScope.experimentLocation
-      } == null) {
-      TaskResult(SUCCEEDED, mapOf(
-        "canaryConfig" to canary.copy(scopes = canary.scopes + deployedClusterScope)
+    val deployedCanaryClusters = mutableMapOf<String, String>()
+    stage.execution.withFirstChildOf("Deploy control cluster") {
+      val location = stage.deployments.control.locations.first()
+      val serverGroup =
+        it.mapTo<DeployServerGroupContext>().deployServerGroups[location]
+          ?.first()
+          ?: throw IllegalStateException("Could not find control server group.")
+
+      deployedCanaryClusters.putAll(mapOf(
+        "controlLocation" to location,
+        "controlServerGroup" to serverGroup
+      ))
+    }
+
+    stage.execution.withFirstChildOf("Deploy experiment cluster") {
+      val location = stage.deployments.experiment.locations.first()
+      val serverGroup  =
+        it.mapTo<DeployServerGroupContext>().deployServerGroups[location]
+          ?.first()
+          ?: throw IllegalStateException("Could not find experiment server group.")
+
+      deployedCanaryClusters.putAll(mapOf(
+        "experimentLocation" to location,
+        "experimentServerGroup" to serverGroup
       ))
-    } else {
-      TaskResult.SUCCEEDED
     }
+
+    return TaskResult(SUCCEEDED, emptyMap<String, Any>(), mapOf(
+      "deployedCanaryClusters" to deployedCanaryClusters
```

Comment:
For canary configs using Stackdriver, we'd like to be able to say right here that the resource type we just deployed is a `gce_instance` or `ec2_instance`. But we'd have to look up the canary config and determine that it's using Stackdriver, which is very ugly. I imagine other metric stores are going to want to do something similar.

Reply:
Could the type of metric service in use be provided in advance via the stage config? Doesn't the pipeline editor already have this info when the canary stage is first configured?